### PR TITLE
grandpa: send neighbor packets to light clients on set transition

### PR DESCRIPTION
**Pull request by andresilva**
As part of the GRANDPA gossip protocol peers exchange small messages (neighbor packets) to inform their peers on what is their local view of the GRANDPA protocol, i.e. what round and set id they are currently observing. This is then used by the gossip logic to avoid gossiping information that peers aren't interested in (either stale or future). Currently these neighbor packets aren't sent to light clients, the reasoning being that since light clients aren't following the full GRANDPA protocol (i.e. they just import commits to finalize blocks), they don't need to care about the state of the peers they are connected to. Still this information can be useful to light clients around validator set boundaries, since it allows light clients to make an informed decision on who to ask justifications from (if necessary), i.e. if I'm looking for a justification for a block that changes the set id from 1 to 2, then any peer that announces being on set id 2 (or higher) should be able to provide such a justification.

This PR changes the gossip logic to also send neighbor packets to light clients when transitioning to a new set. Additionally, the light client also gets a neighbor packet on the initial connection to the peer (this was already the case before this PR).

Related: https://github.com/smol-dot/smoldot/issues/239

### CODE REVIEW

File: client/consensus/grandpa/src/communication/gossip.rs [hunk 1]
```diff
@@ -791,68 +791,72 @@ impl<Block: BlockT> Inner<Block> {
...
 
-			local_view.update_round(round);
+		local_view.update_round(round);
 
-			self.live_topics.push(round, set_id);
-			self.peers.reshuffle();
-		}
-		self.multicast_neighbor_packet()
+		self.live_topics.push(round, set_id);
+		self.peers.reshuffle();
```
Comment (andresilva) [inline client/consensus/grandpa/src/communication/gossip.rs:818]: Nothing changed here, I just removed this useless scope (rustc was dumber when this code was written).

File: client/consensus/grandpa/src/communication/gossip.rs [hunk 2]
```diff
@@ -791,68 +791,72 @@ impl<Block: BlockT> Inner<Block> {
...
+					// such behavior is punishable.
+					return None
+				} else {
+					v
+				},
+		};
+
+		local_view.update_set(set_id);
+		self.live_topics.push(Round(1), set_id);
+		self.authorities = authorities;
```
Comment (andresilva) [inline client/consensus/grandpa/src/communication/gossip.rs:854]: Same as above.

File: client/consensus/grandpa/src/communication/gossip.rs (hunk 1 above)
Comment (andresilva) [inline client/consensus/grandpa/src/communication/gossip.rs:818]: Nothing changed here, I just removed this useless scope (rustc was dumber when this code was written).

File: client/consensus/grandpa/src/communication/gossip.rs (hunk 2 above)
Comment (andresilva) [inline client/consensus/grandpa/src/communication/gossip.rs:854]: Same as above.

File: client/consensus/grandpa/src/communication/gossip.rs [hunk 3]
```diff
@@ -791,68 +791,72 @@ impl<Block: BlockT> Inner<Block> {
...
+			None => return None,
+			Some(ref mut v) =>
+				if v.round == round {
+					// Do not send neighbor packets out if `round` has not changed ---
+					// such behavior is punishable.
+					return None
+				} else {
+					v
+				},
+		};
```
Comment (bkchr) [inline client/consensus/grandpa/src/communication/gossip.rs:804]:
```suggestion
		let local_view = self.local_view.as_mut()?;
		if v.round == round {
			// Do not send neighbor packets out if `round` has not changed ---
			// such behavior is punishable.
			return None
		}
```

File: client/consensus/grandpa/src/communication/gossip.rs [hunk 4]
```diff
@@ -1193,7 +1195,7 @@ impl<Block: BlockT> Inner<Block> {
 		(neighbor_topics, action, catch_up, report)
 	}
 
-	fn multicast_neighbor_packet(&self) -> MaybeMessage<Block> {
+	fn multicast_neighbor_packet(&self, include_light: bool) -> MaybeMessage<Block> {
```
Comment (bkchr) [inline client/consensus/grandpa/src/communication/gossip.rs:1198]:
```suggestion
	fn multicast_neighbor_packet(&self, force_light: bool) -> MaybeMessage<Block> {
```

File: client/consensus/grandpa/src/communication/gossip.rs [hunk 5]
```diff
@@ -1207,8 +1209,9 @@ impl<Block: BlockT> Inner<Block> {
 				.iter()
 				.filter_map(|(id, info)| {
 					// light clients don't participate in the full GRANDPA voter protocol
-					// and therefore don't need to be informed about view updates
-					if info.roles.is_light() {
+					// and therefore don't need to be informed about all view updates unless
+					// we explicitly require it (e.g. when transitioning to a new set)
+					if info.roles.is_light() && !include_light {
```
Comment (bkchr) [inline client/consensus/grandpa/src/communication/gossip.rs:1214]:
```suggestion
					if info.roles.is_light() && !force_light {
```

**Review by bkchr — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]
LGTM!

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2502830

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2502826

**Comment by andresilva** [comment]
bot merge

**Review by melekes — APPROVED** [review]

**Comment by tomaka** [comment]
Is still included in Polkadot v0.9.40?

**Comment by bkchr** [comment]
Is part of 0.9.40

**Pull request merged**